pallets/ethereum-transaction/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+	pallet_prelude::*,
+};
+pub use pallet::*;
+use pallet_evm::{ExitError, ExitFatal, ExitReason};
+use sp_core::{H160, H256, U256};
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+const TRANSACTION_RECOVERY_ID: u64 = 42;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_ethereum::Config {
+		/// The event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	/// Storage for nonce.
+	#[pallet::storage]
+	pub(crate) type Nonce<T: Config> = StorageValue<_, U256, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A call was executed.
+		Executed {
+			from: H160,
+			to: H160,
+			exit_reason: ExitReason,
+			value: Vec<u8>,
+		},
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn get_transaction_signature() -> Option<TransactionSignature> {
+			//TODO(cdamian): Same signature as the one in ethereum-xcm.
+			TransactionSignature::new(
+				TRANSACTION_RECOVERY_ID,
+				H256::from_low_u64_be(1u64),
+				H256::from_low_u64_be(1u64),
+			)
+		}
+	}
+
+	impl<T: Config> EthereumTransactor for Pallet<T> {
+		/// This implementation serves as a wrapper around the Ethereum pallet
+		/// execute functionality. It keeps track of the nonce used for each
+		/// call and builds a fake signature for executing the provided call.
+		///
+		/// NOTE - The execution fees are charged by the Ethereum pallet,
+		/// we only have to charge for the nonce read operation.
+		fn call(
+			from: H160,
+			to: H160,
+			data: &[u8],
+			value: U256,
+			gas_price: U256,
+			gas_limit: U256,
+		) -> DispatchResultWithPostInfo {
+			let nonce = Nonce::<T>::get();
+			let read_weight = T::DbWeight::get().reads(1);
+
+			let signature =
+				Pallet::<T>::get_transaction_signature().ok_or(DispatchErrorWithPostInfo {
+					post_info: PostDispatchInfo {
+						actual_weight: Some(read_weight),
+						pays_fee: Pays::Yes,
+					},
+					error: DispatchError::Other("Failed to create transaction signature"),
+				})?;
+
+			let transaction = TransactionV2::Legacy(LegacyTransaction {
+				nonce,
+				gas_price,
+				gas_limit,
+				action: TransactionAction::Call(to),
+				value,
+				input: data.into(),
+				signature,
+			});
+
+			Nonce::<T>::put(nonce.saturating_add(U256::one()));
+
+			let (_target, _value, info) = pallet_ethereum::Pallet::<T>::execute(
+				from,
+				&transaction,
+				Some(T::config().clone()),
+			)
+			.map_err(|e| {
+				let weight = e.post_info.actual_weight.map_or(Weight::zero(), |w| w);
+
+				DispatchErrorWithPostInfo {
+					post_info: PostDispatchInfo {
+						actual_weight: Some(weight.saturating_add(read_weight)),
+						pays_fee: Pays::Yes,
+					},
+					error: e.error,
+				}
+			})?;
```

Comment:
Because in the case of an error here, nothing will be charged by the ethereum pallet, therefore, we should charge for the weight that is provided by the ethereum pallet in this error, plus our read operation.

Reply:
But it does return a weight if an error is encountered. The idea here is to add our read weight to whatever weight is returned by the Ethereum pallet in the case of a failure.